### PR TITLE
Make Daemons tolerate NoExecute taints correctly

### LINKED ISSUE #45348: DefaultTolerationSeconds admission controller overwrites existing tolerations
At least at think it's what happens.
The code in daemoncontroller says:
```go
// DaemonSet pods shouldn't be deleted by NodeController in case of node problems.
// Add infinite toleration for taint notReady:NoExecute here
// to survive taint-based eviction enforced by NodeController
// when node turns not ready.
v1helper.AddOrUpdateTolerationInPod(newPod, &v1.Toleration{
	Key:      metav1.TaintNodeNotReady,
	Operator: v1.TolerationOpExists,
	Effect:   v1.TaintEffectNoExecute,
})

// DaemonSet pods shouldn't be deleted by NodeController in case of node problems.
// Add infinite toleration for taint unreachable:NoExecute here
// to survive taint-based eviction enforced by NodeController
// when node turns unreachable.
v1helper.AddOrUpdateTolerationInPod(newPod, &v1.Toleration{
	Key:      metav1.TaintNodeUnreachable,
	Operator: v1.TolerationOpExists,
	Effect:   v1.TaintEffectNoExecute,
})
```
And code in admission controller says:
```go
for _, toleration := range tolerations {
	if (toleration.Key == metav1.TaintNodeNotReady || len(toleration.Key) == 0) &&
		(toleration.Effect == api.TaintEffectNoExecute || len(toleration.Effect) == 0) {
		toleratesNodeNotReady = true
	}
	if (toleration.Key == metav1.TaintNodeUnreachable || len(toleration.Key) == 0) &&
		(toleration.Effect == api.TaintEffectNoExecute || len(toleration.Effect) == 0) {
		toleratesNodeUnreachable = true
	}
}
```

Which seems fine. But for some reason I get only 300s toleration on fluentd daemons:
```shell
$kubectl describe pod fluentd-gcp-v2.0-3n134 --namespace=kube-system
...
Tolerations:	node.alpha.kubernetes.io/ismaster=:NoSchedule
		node.alpha.kubernetes.io/notReady=:Exists:NoExecute for 300s
		node.alpha.kubernetes.io/unreachable=:Exists:NoExecute for 300s
...
```

Effect is as expected - daemons get evicted and they're recreated straight away.
@kubernetes/sig-scheduling-bugs

### PR DESCRIPTION
**Pull request by gmarek**
Fix #45348

@kubernetes/sig-scheduling-pr-reviews 

```release-note
Fix a bug that caused invalid Tolerations to be applied to Pods created by DaemonSets.
```

### CODE REVIEW

**Comment by gmarek** [comment]
cc @k82cn

**Comment by gmarek** [comment]
I was considering if we should allow users to specify their own toleration for NodeController taints, but I decided against that, as setting other than this one makes no sense.

File: pkg/api/v1/helper/helpers.go [hunk 1]
```diff
@@ -203,10 +203,10 @@ func NodeSelectorRequirementsAsSelector(nsm []v1.NodeSelectorRequirement) (label
 	return selector, nil
 }
 
-// AddOrUpdateTolerationInPod tries to add a toleration to the pod's toleration list.
+// AddOrUpdateTolerationInPod tries to add a toleration to the toleration list in PodSpec.
```
Comment (k82cn) [inline pkg/api/v1/helper/helpers.go]: s/AddOrUpdateTolerationInPod/AddOrUpdateTolerationInPodSpec

Comment (gmarek) [inline pkg/api/v1/helper/helpers.go]: Done.

**Review by k82cn — APPROVED** [review]
LGTM :).

**Comment by gmarek** [comment]
We should cherry-pick it (or other fix if we decide to) to the next patch release. FYI @ethernetdan @enisoc

**Comment by davidopp** [comment]
Do you need to delete the code here?
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/daemoncontroller.go#L858

**Comment by k82cn** [comment]
> Do you need to delete the code here?
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/daemoncontroller.go#L858

@davidopp , I think we need to keep this because of Tolerant Predicates.

**Comment by k82cn** [comment]
@k8s-bot unit test this

**Comment by gmarek** [comment]
@davidopp - yup, exactly as @k82cn - I believe that if Daemons tolarate those taints they should be able to schedule on tainted nodes as well.

**Comment by gmarek** [comment]
@k8s-bot bazel test this

**Comment by krzyzacy** [comment]
ref https://github.com/kubernetes/test-infra/issues/2666

**Comment by gmarek** [comment]
@k8s-bot bazel test this

**Pull request merged**